### ISSUE TITLE
Youtube-dl don't work

### ISSUE THREAD

**Issue by marcbth** [ArturSierzant/OMPD#110]
Hello Artur,

tank for your update, but

when I install youtube-dl

sudo curl -L https://yt-dl.org/downloads/latest/youtube-dl -o /usr/local/bin/youtube-dl
sudo chmod a+rx /usr/local/bin/youtube-dl

and

$cfg['python_path'] = '/usr/bin/python3.5';
$cfg['youtube-dl_path'] = '/usr/local/bin/youtube-dl';

youtube-dl doesn't work.

can you help me

thanks

cliff

**Comment by ArturSierzant** [comment]
Try this:
```
$cfg['python_path'] = '';
$cfg['youtube-dl_path'] = '/usr/local/bin/youtube-dl';
```
When O!MPD gets info from YouTube, it calls `youtube-dl` by running command:
`$cfg['python_path']  $cfg['youtube-dl_path'] $cfg['youtube-dl_options'] [YouTube_url]`
so if you have binary version of `youtube-dl`, path to python is not needed. On the other hand, when you use python-version of `youtube-dl`, you might need `$cfg['python_path']` properrly defined. 

Anyway, in your case, empty `$cfg['python_path']` should solve the problem.

Regards,
Artur

**Comment by marcbth** [comment]
Thank you.
I'll try it next year.

Cliff

**Comment by marcbth** [comment]
Hello Artur,

$cfg['python_path'] = '';
$cfg['youtube-dl_path'] = '/usr/local/bin/youtube-dl';

dont work.

i have debian buster(amd64) , python 3.7.

when searching the youtube files are displayed. but only one wheel which is turning. the same happens when i want to use "add

can you help me?


Many thanks

Cliff

**Comment by marcbth** [comment]
javascript:ajaxRequest('play.php?action=insertSelect&playAfterInsert=yes&track_id=youtube_TvnYmWpD_T8&positon_id=50001',evaluateAdd);

not running

**Comment by ArturSierzant** [comment]
You can see YT results, because search is done by parsing html page - `youtube-dl` is not used here. It is used for getting url of specified audio stream and this part is not working in your case. There may be some errors in script execution  - please try to look into your www/php server logs.
But first of all check, if `youtube-dl` works properly: try to run the following command:

` /usr/local/bin/youtube-dl -j --no-check-certificate --prefer-insecure TvnYmWpD_T8`

In return you should see json-encoded  text with a lot of information concerning this video ("Prince - Purple Rain (Official Video)").

Also check if you have those 2 options set as following:
```
$cfg['youtube-dl_options'] = '-j --no-check-certificate --prefer-insecure';
$cfg['youtube_audio_format_name'] = '140 - audio only (tiny)';
```

Regards,
Artur

**Comment by marcbth** [comment]
Hello Artur, 

thank you very much for your answer.

i'm getting an issue:
/usr/local/bin/youtube-dl -j --no-check-certificate --prefer-insecure TvnYmWpD_T8

but the problem persists.

config:

unset($cfg['youtube_indicator']);

$cfg['python_path'] = '';
$cfg['youtube-dl_path'] = '/usr/local/bin/youtube-dl';
$cfg['youtube-dl_options'] = '-j --no-check-certificate --prefer-insecure';
$cfg['youtube_audio_format_name'] = '140 - audio only (tiny)';
$cfg['youtube_indicator'][] = 'www.youtube.';
$cfg['youtube_indicator'][] = 'youtu.be/';
$cfg['youtube_indicator'][] = 'm.youtube.';

Cliff
![picture](https://user-images.githubusercontent.com/23519079/71975002-4153cd80-3213-11ea-81c0-57c309ae5ded.png)

**Comment by marcbth** [comment]
without -j

/usr/local/bin/youtube-dl  --no-check-certificate --prefer-insecure -f '(bestvideo+bestaudio/best)[protocol^=http]' TvnYmWpD_T8

[youtube] TvnYmWpD_T8: Downloading webpage
[youtube] TvnYmWpD_T8: Downloading video info webpage
WARNING: You have requested multiple formats but ffmpeg or avconv are not installed. The formats won't be merged.
WARNING: Requested formats are incompatible for merge and will be merged into mkv.
[download] Destination: Prince - Purple Rain (Official Video)-TvnYmWpD_T8.f398.mp4
[download] 100% of 37.18MiB in 00:17
[download] Destination: Prince - Purple Rain (Official Video)-TvnYmWpD_T8.f251.webm
[download] 100% of 7.34MiB in 00:07

**Comment by marcbth** [comment]
only audio 

/usr/local/bin/youtube-dl --no-check-certificate --prefer-insecure -f '(bestaudio/best)[protocol^=http]' TvnYmWpD_T8
[youtube] TvnYmWpD_T8: Downloading webpage
[youtube] TvnYmWpD_T8: Downloading video info webpage
[download] Destination: Prince - Purple Rain (Official Video)-TvnYmWpD_T8.webm
[download] 100% of 7.34MiB in 00:03

**Comment by marcbth** [comment]
Hello Artur
i have taken an ompd version from november and youtube add works there.

Could it be the htlmphp parser?

Cliff

**Comment by marcbth** [comment]
play.php line 1124 ....

its new. Its a problem?

**Comment by ArturSierzant** [comment]
So 
`/usr/local/bin/youtube-dl -j --no-check-certificate --prefer-insecure TvnYmWpD_T8`
gives you an error? Do I get it right? Or rather you receive something that starts like this:
```
{"upload_date": "20160422", "extractor": "youtube", "series": null, "format": "398 - 1280x720 (720p)+251 - audio only (tiny)", "vbr": null, "chapters": null, "height": 720, "like_count": 736922, "duration": 477, "fulltitle": "Prince - Purple Rain (Official Video)", "playlist_index": null, "album": "Purple Rain", "view_count": 156139067, "playlist": null, "title": "Prince - Purple Rain (Official Video)", "_filename": "Prince - Purple Rain (Official Video)-TvnYmWpD_T8.mp4", "creator": "Prince & The Revolution", "ext": "mp4", "id": "TvnYmWpD_T8", "dislike_count": 44311, "average_rating": 4.7731228, "abr": 160, "uploader_url": "http://www.youtube.com/channel/UCPzbKgAJ6gj0xMVgMHIV2gQ", "categories": ["People & Blogs"], "fps": 30, "stretched_ratio": null, "season_number": null, "annotations": null, "webpage_url_basename": "TvnYmWpD_T8", "acodec": "opus", "display_id": "TvnYmWpD_T8", "requested_formats": [{"asr": null, "tbr": 972.707, "protocol": "https", "format": "398 - 1280x720 (720p)", "url":
```
The above command must work with ` -j` - in the other case `youtube-dl` will download file instead of json data (O!MPD needs json data to retrieve needed data).  


When you click add button, the following request is send to server:
`http://192.168.1.120/ompd/play.php?action=addSelect&track_id=youtube_jdyto5rf0HU&ajax=1` 
Try to paste it to your web browser (change IP/app dir) and see the answer, which should look like this:
```
{"addResult":"add_OK","album_id":null,"track_id":"youtube_jdyto5rf0HU","disc":null,"favorite_id":null,"random":null}
```

Regards,
Artur

**Comment by marcbth** [comment]
Hello Artur,

thank you very much for your trouble.

when I type /usr/local/bin/youtube-dl -j --no-check-certificate --prefer-insecure TvnYmWpD_T8

the answer is
{"id": "TvnYmWpD_T8", "uploader": "The Codfather", "uploader_id": "UCPzbKgAJ6gj0xMVgMHIV2gQ", "uploader_url": "http://www.youtube.com/channel/UCPzbKgAJ6gj0xMVgMHIV2gQ", "channel_id": "UCPzbKgAJ6gj0xMVgMHIV2gQ", "channel_url": "http://www.youtube.com/channel/UCPzbKgAJ6gj0xMVgMHIV2gQ", "upload_date": "20160422", "license": null, "creator": "Prince & The Revolution", "title": "Prince - Purple Rain (Official Video)", "alt_title": "Purple Rain", "thumbnail": "https://i.ytimg.com/vi/TvnYmWpD_T8/maxresdefault.jpg", 

when I type

http://192.168.178.106/ompd/play.php?action=addSelect&track_id=youtube_jdyto5rf0HU&ajax=1

better for me

http://192.168.178.106/play.php?action=addSelect&track_id=youtube_jdyto5rf0HU&ajax=1

then 
192.168.178.106 kann diese Anfrage momentan nicht verarbeiten.
HTTP ERROR 500

what surprises me is that it works with an older version of ompd. at "now playing" i can "add" a youtube link and play

Cliff

Mogę wysłać ci mój identyfikator Teamviewer pocztą prywatną.

**Comment by marcbth** [comment]
in the old version from ompd

http://192.168.178.106/play.php?action=addSelect&track_id=youtube_jdyto5rf0HU&ajax=1

gives

{"addResult":"add_OK","album_id":null,"track_id":"youtube_jdyto5rf0HU","disc":null,"favorite_id":null,"random":null}

**Comment by ArturSierzant** [comment]
OK, looks like `youtube-dl` works OK (the more so because it also works in older version O!MPD).

I think it's a problem in `getYoutubeStreamUrl` function. Try this: in O!MPD root dir create file, say, `testYT.php` and paste into it the code below: 
```
<?php
require_once('include/initialize.inc.php');
require_once('include/library.inc.php');
global $cfg, $db;
@ini_set('display_errors', 'On');
echo getYoutubeStreamUrl('TvnYmWpD_T8','');
?>
```
Then run it:
http://192.168.178.106/testYT.php
Let's see the answer.

Regards,
Artur

**Comment by marcbth** [comment]
Hello Artur,

the result

https://r4---sn-8xgn5uxa-i5he.googlevideo.com/videoplayback?expire=1578631642&ei=eq0XXo3cD52c1gLbgZa4CA&ip=88.130.88.100&id=o-AOmWGwke5o_lgTCsGF5tkiCmWbSL80vU6_qiiIOScWZf&itag=140&source=youtube&requiressl=yes&mm=31%2C29&mn=sn-8xgn5uxa-i5he%2Csn-i5h7lner&ms=au%2Crdu&mv=m&mvi=3&pcm2cms=yes&pl=20&initcwndbps=1427500&vprv=1&mime=audio%2Fmp4&gir=yes&clen=7723968&dur=477.216&lmt=1569827527989628&mt=1578609884&fvip=4&keepalive=yes&fexp=23842630&c=WEB&txp=5531432&sparams=expire%2Cei%2Cip%2Cid%2Citag%2Csource%2Crequiressl%2Cvprv%2Cmime%2Cgir%2Cclen%2Cdur%2Clmt&lsparams=mm%2Cmn%2Cms%2Cmv%2Cmvi%2Cpcm2cms%2Cpl%2Cinitcwndbps&lsig=AHylml4wRgIhAOFi_79vn-dxo-YVw0JhsE3ghEA5GpksXBLAI9l165EkAiEA39JrKGJQgOqjmYV_SaZjXMMKdDkJ3F5y_NNILeytTDI%3D&sig=ALgxI2wwRgIhAJYr6NEEkzgwELh4kO93rMs_8PgAzAj-MOFGWd5m2xkZAiEA1jCjrtOfyOsB680kk-xWoI94W_8Q05rwwsVHTEezRDk=&ratebypass=yes&ompd_title=Purple+Rain+%28Official+Video%29&ompd_duration=477&ompd_artist=Prince+%26+The+Revolution&ompd_thumbnail=https%3A%2F%2Fi.ytimg.com%2Fvi%2FTvnYmWpD_T8%2Fmaxresdefault.jpg&ompd_year=2016&ompd_webpage=http%3A%2F%2Fwww.youtube.com%2Fwatch%3Fv%3DTvnYmWpD_T8

**Comment by marcbth** [comment]
Thanks again for your effort

**Comment by ArturSierzant** [comment]
I think I've found the problem - it's missing argument in `getYoutubeStreamUrl` function. To fix it go to:
https://github.com/ArturSierzant/OMPD/blob/dd1dd902d90b7c1769db621e533656e1db276c8d/include/library.inc.php#L1249
and change it to:
`function getYoutubeStreamUrl($url, $title = ''){`

It's strange - it works OK with my RaspberryPi (PHP 7.0.33) but it's not working with my Win (PHP 7.4.1)...

I'll fix it in next commit, as soon as I finish what I'm working on now.

Regards,
Artur

**Comment by marcbth** [comment]
Perfect. It was.

It's on now.

Thank you so much.

Cliff

Oh, yeah, I saw the Radio Staion Covers folder. Will there be a function at the top of the menu bar called "Radio"? Right now I'm working on favorites. Not Ideal. I'm going to open up a new Issues.

Oh yeah, I've got a new dark skin with 1980px, so it's full width. Put it online for everyone to use.

**Comment by marcbth** [comment]
if you are interested I can send you an Odriod H2

Cliff

**Comment by ArturSierzant** [comment]
I wrote this a time ago, but (I don't know why) I didn't published it... Anyway I fixed it in 1dd93cd

---------------------------- 

I think I've found the problem - it's missing argument in `getYoutubeStreamUrl` function. To fix it go to:
https://github.com/ArturSierzant/OMPD/blob/dd1dd902d90b7c1769db621e533656e1db276c8d/include/library.inc.php#L1249
and change it to:
`function getYoutubeStreamUrl($url, $title = ''){`

It's strange - it works OK with my RaspberryPi (PHP 7.0.33) but it's not working with my Win (PHP 7.4.1)...

I'll fix it in next commit, as soon as I finish what I'm working on now.

Regards,
Artur
